Added a Backend validation for not allowing users to submit questions with empty custom choice options

hrit2773 opened the pull request:
## Proposed Changes

- Added a backend validation to check if all the answer options has a value in it. Even if one option has value field empty it will raise a value Error

### Associated Issue

- Fixes https://github.com/ohcnetwork/care_fe/issues/11332
- Fixes https://github.com/ohcnetwork/care/issues/2930

##Screen Recording

https://github.com/user-attachments/assets/400cbb23-adea-4ff8-81f1-7a6b5b3e51a9



## Merge Checklist

- [x] Tests added/fixed
- [ ] Update docs in `/docs`
- [ ] Linting Complete
- [x] Screen recording added

_*Only PR's with test cases included and passing lint and test pipelines will be reviewed*_

@Jacobjeevan @rithviknishad @gigincg @vigneshhari 


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **New Features**
  - Introduced validation for answer option values in questionnaires, ensuring that custom choices are not left blank and providing clear error messages for invalid submissions.

- **Tests**
  - Added tests to confirm that invalid answer option values result in appropriate error responses during questionnaire submissions.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

hrit2773 committed:
Added a validation for not allowing users to submit questions with empty custom options

hrit2773 committed:
Merge branch 'develop' into new

hrit2773 commented:
@Jacobjeevan @rithviknishad can I get a review on this. Not sure if test cases are required

hrit2773 committed:
changed to all() for better readability

hrit2773 committed:
fixed accidental change

vigneshhari commented:
@DraKen0009 review required

vigneshhari commented:
@hrit2773 please create an issue and get it assigned before working on the issue.

sainak commented:
looks good, but tests missing

hrit2773 commented:
> @hrit2773 please create an issue and get it assigned before working on the issue.

It was linked to an issue in care_fe repository https://github.com/ohcnetwork/care_fe/issues/11332 so l linked this PR to that

hrit2773 commented:
@sainak Will update the tests

praffq commented on care/emr/resources/questionnaire/spec.py:
```diff
@@ -171,6 +171,13 @@ def validate_value_set_or_options(self):
             raise ValueError(err)
+
+        if self.answer_option:
+            options_present = all(option.value for option in self.answer_option)
```
~~This will give error if the value is 0. 
Check specifically for None or empty string, whichever is sent by the frontend~~

0 is sent as "0". so this change should work

praffq replied:
@hrit2773  use Falsey values in test cases though, just in case to check if it works fine

hrit2773 replied:
@DraKen0009 option.value will return True if there is a string as the values sent are in the form of string. so the user cant sent '' empty string. The value is anyways in the form of string so no issues there.

hrit2773 replied:
@DraKen0009 Falsey values like?? can you give some example falsey values will add them in tests. Any string which is non empty will work fine.

hrit2773 committed:
Merge branch 'develop' into new

hrit2773 committed:
added test for false condition and modified the string to remove leading and trailing spaces

hrit2773 committed:
fixed back

hrit2773 commented:
@DraKen0009 @sainak I have updated the test to add a faulty value : "  " and changed to option.value.strip() so that the users cannot add any leading or trailing spaces. the tests are passing. Can you guys review it again.

vigneshhari commented:
> > @hrit2773 please create an issue and get it assigned before working on the issue.
> 
> It was linked to an issue in care_fe repository [ohcnetwork/care_fe#11332](https://github.com/ohcnetwork/care_fe/issues/11332) so l linked this PR to that

Backend contributions only after an issue is opened in the BE repo and assigned.

hrit2773 commented:
> > > @hrit2773 please create an issue and get it assigned before working on the issue.
> > 
> > 
> > It was linked to an issue in care_fe repository [ohcnetwork/care_fe#11332](https://github.com/ohcnetwork/care_fe/issues/11332) so l linked this PR to that
> 
> Backend contributions only after an issue is opened in the BE repo and assigned.

@vigneshhari I have opened an issue for this do check it out and assign me.

sainak commented on care/emr/resources/questionnaire/spec.py:
```diff
@@ -171,6 +171,13 @@ def validate_value_set_or_options(self):
             err = "Either answer options or a value set must be provided for choice type questions"
             raise ValueError(err)
+
+        if self.answer_option:
+            options_present = all(option.value.strip() for option in self.answer_option)
+            if not options_present:
+                err="All the answer options must be provided for custom choices"
+                raise ValueError(err)
+
```
it would be better to move this validation to the AnswerOption spec

hrit2773 replied:
@sainak yes we can do that will update it

hrit2773 replied:
@sainak pushed the changes you can check it

hrit2773 committed:
shifted the test to AnswerSpec

hrit2773 commented:
@sainak the values should not repeat also right? should I add validation for that as well?

vigneshhari commented on care/emr/resources/questionnaire/spec.py:
```diff
@@ -95,6 +95,12 @@ class AnswerOption(QuestionnaireBaseSpec):
     )
+    @field_validator("value")
+    @classmethod
+    def validate_title(cls, value: str, info):
```
Better function name please.

hrit2773 replied:
@vigneshhari changed

hrit2773 committed:
fixed accidental name change

hrit2773 committed:
Merge branch 'develop' into new

hrit2773 commented:
@vigneshhari @sainak @DraKen0009 can I get a final review on this

praffq commented:
Looks good !

hrit2773 commented:
@sainak @vigneshhari your reviews??

praffq committed:
Merge branch 'develop' into new

praffq commented:
@hrit2773  lint error

hrit2773 committed:
lint fix

hrit2773 commented:
> @hrit2773 lint error

fixed

hrit2773 commented:
@sainak can you re run it again
